src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -1436,15 +1540,29 @@ public bool CleanCacheAndDisposeIfUnused()
                 // Get the current time.  This is compared against each connection's last returned
                 // time to determine whether a connection is too old and should be closed.
                 long nowTicks = Environment.TickCount64;
-                Http2Connection? http2Connection = _http2Connection;
+                // Copy the reference to a local variable to simplify the removal logic below.
+                Http2Connection[]? localHttp2Connections = _http2Connections;
 
-                if (http2Connection != null)
+                if (localHttp2Connections != null)
                 {
-                    if (http2Connection.IsExpired(nowTicks, pooledConnectionLifetime, pooledConnectionIdleTimeout))
+                    for (int i = 0; i < localHttp2Connections.Length; i++)
                     {
-                        http2Connection.Dispose();
-                        // We can set _http2Connection directly while holding lock instead of calling InvalidateHttp2Connection().
-                        _http2Connection = null;
+                        Http2Connection http2Connection = localHttp2Connections[i];
+                        if (http2Connection != null)
+                        {
+                            if (http2Connection.IsExpired(nowTicks, pooledConnectionLifetime, pooledConnectionIdleTimeout))
+                            {
+                                http2Connection.Dispose();
+                                if (_http2Connections != null)
+                                {
+                                    if (TryRemoveHttp2Connection(http2Connection, _http2Connections, out Http2Connection[] newConnections))
+                                    {
+                                        // We can set _http2Connections directly while holding lock instead of calling InvalidateHttp2Connection().
```

Comment:
We are holding the lock here, so the remove should never fail, right? If so we should assert this.

Reply:
Debug.Fail added.